replica/distributed_loader.cc

```diff
@@ -448,57 +428,151 @@ future<> distributed_loader::handle_sstables_pending_delete(sstring pending_dele
     co_await when_all_succeed(futures.begin(), futures.end()).discard_result();
 }
 
-future<> distributed_loader::populate_column_family(distributed<replica::database>& db, sstring sstdir, sstring ks, sstring cf, allow_offstrategy_compaction do_allow_offstrategy_compaction, must_exist dir_must_exist) {
-    dblog.debug("Populating {}/{}/{} allow_offstrategy_compaction={} must_exist={}", ks, cf, sstdir, do_allow_offstrategy_compaction, dir_must_exist);
-    return async([&db, sstdir = std::move(sstdir), ks = std::move(ks), cf = std::move(cf), do_allow_offstrategy_compaction, dir_must_exist] {
+class sstables_metadata {
+    distributed<replica::database>& _db;
+    sstring _ks;
+    sstring _cf;
+    global_column_family_ptr _global_table;
+    std::unordered_map<sstring, lw_shared_ptr<sharded<sstables::sstable_directory>>> _directories;
+    sstables::sstable::version_types _highest_version;
+    sstables::generation_type _highest_generation = sstables::generation_from_value(0);
+    fs::path _base_path;
+
+public:
+    sstables_metadata(distributed<replica::database>& db, sstring ks, sstring cf)
+        : _db(db)
+        , _ks(std::move(ks))
+        , _cf(std::move(cf))
+        , _global_table(_db, _ks, _cf)
+        , _base_path(_global_table->dir())
+    {}
+
+    ~sstables_metadata() {
+        // All directories must have been stopped
+        // using sstables_metadata::stop()
+        assert(_directories.empty());
+    }
+
+    fs::path get_path(sstring subdir = "") const {
+        return subdir.empty() ? _base_path : _base_path / subdir;
+    }
+
+    future<> start() {
         assert(this_shard_id() == 0);
 
-        if (!file_exists(sstdir).get0()) {
-            if (dir_must_exist) {
-                throw std::runtime_error(format("Populating {}/{} failed: {} does not exist", ks, cf, sstdir));
+        for (auto subdir : { "", sstables::staging_dir, sstables::quarantine_dir }) {
+            auto sstpath = get_path(subdir);
```

Comment:
perhaps we should have start_directory(sstring subdir), where all initialization code goes into it. therefore indentation is kept and start() will simply iterate through subdirs, and call start_directory(...) on each. simplifies reviewing and makes the code easier to follow.

Reply:
okay